source/dotnet/Library/AdaptiveCards.Rendering.Html/AdaptiveRenderContext.cs

```diff
@@ -18,31 +18,81 @@ public AdaptiveRenderContext(AdaptiveHostConfig hostConfig, AdaptiveElementRende
 
         public AdaptiveElementRenderers<HtmlTag, AdaptiveRenderContext> ElementRenderers { get; set; }
 
+        public AdaptiveFeatureRegistration FeatureRegistration { get; set; }
+
         public IList<AdaptiveWarning> Warnings { get; } = new List<AdaptiveWarning>();
 
         public IList<HtmlTag> ShowCardTags { get; } = new List<HtmlTag>();
 
+        private bool AncestorHasFallback = false;
+
         public HtmlTag Render(AdaptiveTypedElement element)
         {
-            // If non-inertactive, inputs should just render text
-            if (!Config.SupportsInteractivity && element is AdaptiveInput input)
+            HtmlTag htmlTagOut = null;
+            var oldAncestorHasFallback = AncestorHasFallback;
+            var elementHasFallback = element != null && element.Fallback != null && (element.Fallback.Type != AdaptiveFallbackElement.AdaptiveFallbackType.None);
+            AncestorHasFallback = AncestorHasFallback || elementHasFallback;
+
+            try
             {
-                var tb = new AdaptiveTextBlock();
-                tb.Text = input.GetNonInteractiveValue();
-                Warnings.Add(new AdaptiveWarning(-1, $"Rendering non-interactive input element '{element.Type}'"));
-                return Render(tb);
-            }
+                if (AncestorHasFallback && !element.MeetsRequirements(FeatureRegistration))
+                {
+                    throw new AdaptiveFallbackException("Element requirements aren't met");
+                }
 
-            var renderer = ElementRenderers.Get(element.GetType());
-            if (renderer != null)
+                // If non-interactive, inputs should just render text
+                if (!Config.SupportsInteractivity && element is AdaptiveInput input)
+                {
+                    var tb = new AdaptiveTextBlock();
+                    tb.Text = input.GetNonInteractiveValue();
+                    Warnings.Add(new AdaptiveWarning(-1, $"Rendering non-interactive input element '{element.Type}'"));
+                    htmlTagOut = Render(tb);
+                }
+
+                if (htmlTagOut == null)
+                {
+                    var renderer = ElementRenderers.Get(element.GetType());
+                    if (renderer != null)
+                    {
+                        htmlTagOut = renderer.Invoke(element, this);
+                    }
+                }
+            }
+            catch (AdaptiveFallbackException e)
             {
-                return renderer.Invoke(element, this);
+                if (!elementHasFallback)
+                {
+                    throw e;
```

Comment:
This should be just throw to retain the original stack context in the exception that gets bubbled back up (throw e will reset the stack)